File: Subsets 01 Recursion.java
// Time Complexity : O(2^n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    List<List<Integer>> result;
    public List<List<Integer>> subsets(int[] nums) {
        result = new ArrayList<>();
        if(nums == null || nums.length == 0) return result;
        helper(nums, 0 , new ArrayList<>());
        return result;
    }
    
    private void helper(int[] nums, int index, List<Integer> path){
        // base
        if(index == nums.length){
            result.add(path);
            return;
        }
        
        // logic
        // choose
        List<Integer> temp = new ArrayList<>(path);
        temp.add(nums[index]);
        helper(nums, index + 1, temp);
        
        // not choose
        helper(nums, index + 1, new ArrayList<>(path));
    
    }
}

File: Subsets0-1Recursion.py
'''
TC: O(N*2^N) - to generate all subsets and create deep copy for every subset created. 
SC: O(N*2^N)
'''
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        #0/1 Recursion
        self.res = []

        def dfs(idx, path):
            #Base
            if idx == len(nums):
                self.res.append(list(path))
                return

            #Logic
            #Don't Choose
            dfs(idx+1, path)
            
            #Choose
            path.append(nums[idx])
            dfs(idx+1, path)
            path.pop()
        
        dfs(0, [])
        return self.res
s = Solution()
print(s.subsets([1,2,3]))
print(s.subsets([0]))